File: ios/kwift/shared/ObservableProperty.swift
import Foundation
//package com.lightningkite.kwift.shared


//import com.lightningkite.kwift.actuals.AnyObject
//import com.lightningkite.kwift.actuals.escaping
//import com.lightningkite.kwift.actuals.weak
//import com.lightningkite.kwift.actuals.weakLambda

final public class ObservableProperty<T> {
    
    public var underlyingValue: T
    
    init( underlyingValue: T) {
        self.underlyingValue = underlyingValue
    }
    convenience init(
        _ underlyingValue: T
    ){ 
        self.init(
            underlyingValue: underlyingValue
        ) 
    }
    
    var onChange = Event<T>()
    var value: T {
        
        get { return underlyingValue }
        set(value) {
            underlyingValue = value
            onChange.invokeAll(value : value)
        }
    }
    
    
    
    func addAndRunWeak<A : AnyObject>(
        referenceA: A,
        listener: @escaping (A, T) -> Void
    ) -> Subscription<T> { return onChange.addAndRunWeak(
        referenceA : referenceA,
        value : value,
        listener : listener
    ) }
    func addAndRunWeak<A : AnyObject>(
        _ referenceA: A,
        _ listener: @escaping (A, T) -> Void
    ) -> Subscription<T> { return addAndRunWeak(referenceA: referenceA, listener: listener) }
    
    func addAndRunWeak<A : AnyObject, B : AnyObject>(
        referenceA: A,
        referenceB: B,
        listener: @escaping (A, B, T) -> Void
    ) -> Subscription<T> { return onChange.addAndRunWeak(
        referenceA : referenceA,
        referenceB : referenceB,
        value : value,
        listener : listener
    ) }
    func addAndRunWeak<A : AnyObject, B : AnyObject>(
        _ referenceA: A,
        _ referenceB: B,
        _ listener: @escaping (A, B, T) -> Void
    ) -> Subscription<T> { return addAndRunWeak(referenceA: referenceA, referenceB: referenceB, listener: listener) }
    
    func addAndRunWeak<A : AnyObject, B : AnyObject, C : AnyObject>(
        referenceA: A,
        referenceB: B,
        referenceC: C,
        listener: @escaping (A, B, C, T) -> Void
    ) -> Subscription<T> { return onChange.addAndRunWeak(
        referenceA : referenceA,
        referenceB : referenceB,
        referenceC : referenceC,
        value : value,
        listener : listener
    ) }
    func addAndRunWeak<A : AnyObject, B : AnyObject, C : AnyObject>(
        _ referenceA: A,
        _ referenceB: B,
        _ referenceC: C,
        _ listener: @escaping (A, B, C, T) -> Void
    ) -> Subscription<T> { return addAndRunWeak(referenceA: referenceA, referenceB: referenceB, referenceC: referenceC, listener: listener) }
    
    
    func calculatedBy<A>(
        a: ObservableProperty<A>,
        calculation: @escaping (A) -> T
    ) -> ObservableProperty<T> {
        self.value = calculation(a.value)
        a.onChange.add(listener : { [weak self] (aValue) in
            if let self = self {
                self.value = calculation(aValue)
                
            }
        })
        return self
    }
    func calculatedBy<A>(
        _ a: ObservableProperty<A>,
        _ calculation: @escaping (A) -> T
    ) -> ObservableProperty<T> { return calculatedBy(a: a, calculation: calculation) }
    
    func calculatedBy<A, B>(
        a: ObservableProperty<A>,
        b: ObservableProperty<B>,
        calculation: @escaping (A, B) -> T
    ) -> ObservableProperty<T> {
        self.value = calculation(a.value, b.value)
        weak var aWeak = a
        weak var bWeak = b
        a.onChange.add(listener : { [weak self] (aValue) in
            if let self = self {
                var bValue = bWeak?.value;
                if let bValue = bValue {
                    self.value = calculation(aValue, bValue)
                }
                
            }
        })
        b.onChange.add(listener : { [weak self] (bValue) in
            if let self = self {
                var aValue = aWeak?.value;
                if let aValue = aValue {
                    self.value = calculation(aValue, bValue)
                }
                
            }
        })
        return self
    }
    func calculatedBy<A, B>(
        _ a: ObservableProperty<A>,
        _ b: ObservableProperty<B>,
        _ calculation: @escaping (A, B) -> T
    ) -> ObservableProperty<T> { return calculatedBy(a: a, b: b, calculation: calculation) }
    
    func calculatedBy<A, B, C>(
        a: ObservableProperty<A>,
        b: ObservableProperty<B>,
        c: ObservableProperty<C>,
        calculation: @escaping (A, B, C) -> T
    ) -> ObservableProperty<T> {
        self.value = calculation(a.value, b.value, c.value)
        weak var aWeak = a
        weak var bWeak = b
        weak var cWeak = c
        a.onChange.add(listener : { [weak self] (aValue) in
            if let self = self {
                var bValue = bWeak?.value;
                var cValue = cWeak?.value;
                if let bValue = bValue {
                    if let cValue = cValue {
                        self.value = calculation(aValue, bValue, cValue)
                    }
                }
                
            }
        })
        b.onChange.add(listener : { [weak self] (bValue) in
            if let self = self {
                var aValue = aWeak?.value;
                var cValue = cWeak?.value;
                if let aValue = aValue, let cValue = cValue {
                    self.value = calculation(aValue, bValue, cValue)
                }
                
            }
        })
        c.onChange.add(listener : { [weak self] (cValue) in
            if let self = self {
                var aValue = aWeak?.value;
                var bValue = bWeak?.value;
                if let aValue = aValue, let bValue = bValue {
                    self.value = calculation(aValue, bValue, cValue)
                }
                
            }
        })
        return self
    }
    func calculatedBy<A, B, C>(
        _ a: ObservableProperty<A>,
        _ b: ObservableProperty<B>,
        _ c: ObservableProperty<C>,
        _ calculation: @escaping (A, B, C) -> T
    ) -> ObservableProperty<T> { return calculatedBy(a: a, b: b, c: c, calculation: calculation) }
    
    func calculatedBy<A, B, C, D>(
        a: ObservableProperty<A>,
        b: ObservableProperty<B>,
        c: ObservableProperty<C>,
        d: ObservableProperty<D>,
        calculation: @escaping (A, B, C, D) -> T
    ) -> ObservableProperty<T> {
        self.value = calculation(a.value, b.value, c.value, d.value)
        weak var aWeak = a
        weak var bWeak = b
        weak var cWeak = c
        weak var dWeak = d
        a.onChange.add(listener : { [weak self] (aValue) in
            if let self = self {
                var bValue = bWeak?.value;
                var cValue = cWeak?.value;
                var dValue = dWeak?.value;
                if let bValue = bValue, let cValue = cValue, let dValue = dValue {
                    self.value = calculation(aValue, bValue, cValue, dValue)
                }
                
            }
        })
        b.onChange.add(listener : { [weak self] (bValue) in
            if let self = self {
                var aValue = aWeak?.value;
                var cValue = cWeak?.value;
                var dValue = dWeak?.value;
                if let aValue = aValue, let cValue = cValue, let dValue = dValue {
                    self.value = calculation(aValue, bValue, cValue, dValue)
                }
                
            }
        })
        c.onChange.add(listener : { [weak self] (cValue) in
            if let self = self {
                var aValue = aWeak?.value;
                var bValue = bWeak?.value;
                var dValue = dWeak?.value;
                if let aValue = aValue, let bValue = bValue, let dValue = dValue {
                    self.value = calculation(aValue, bValue, cValue, dValue)
                }
                
            }
        })
        d.onChange.add(listener : { [weak self] (dValue) in
            if let self = self {
                var aValue = aWeak?.value;
                var bValue = bWeak?.value;
                var cValue = cWeak?.value;
                if let aValue = aValue, let bValue = bValue, let cValue = cValue {
                    self.value = calculation(aValue, bValue, cValue, dValue)
                }
                
            }
        })
        return self
    }
    func calculatedBy<A, B, C, D>(
        _ a: ObservableProperty<A>,
        _ b: ObservableProperty<B>,
        _ c: ObservableProperty<C>,
        _ d: ObservableProperty<D>,
        _ calculation: @escaping (A, B, C, D) -> T
    ) -> ObservableProperty<T> { return calculatedBy(a: a, b: b, c: c, d: d, calculation: calculation) }
}

File: android/app/src/main/java/com/lightningkite/kwift/shared/ObservableProperty.kt
package com.lightningkite.kwift.shared

import com.lightningkite.kwift.actuals.AnyObject
import com.lightningkite.kwift.actuals.escaping
import com.lightningkite.kwift.actuals.weak
import com.lightningkite.kwift.actuals.weakLambda

class ObservableProperty<T>(var underlyingValue: T) {
    val onChange = Event<T>()
    var value: T
        get() = underlyingValue
        set(value) {
            underlyingValue = value
            onChange.invokeAll(value = value)
        }


    fun <A : AnyObject> addAndRunWeak(
        referenceA: A,
        @escaping listener: (A, T) -> Unit
    ): Subscription<T> = onChange.addAndRunWeak(
        referenceA = referenceA,
        value = value,
        listener = listener
    )

    fun <A : AnyObject, B : AnyObject> addAndRunWeak(
        referenceA: A,
        referenceB: B,
        @escaping listener: (A, B, T) -> Unit
    ): Subscription<T> = onChange.addAndRunWeak(
        referenceA = referenceA,
        referenceB = referenceB,
        value = value,
        listener = listener
    )

    fun <A : AnyObject, B : AnyObject, C : AnyObject> addAndRunWeak(
        referenceA: A,
        referenceB: B,
        referenceC: C,
        @escaping listener: (A, B, C, T) -> Unit
    ): Subscription<T> = onChange.addAndRunWeak(
        referenceA = referenceA,
        referenceB = referenceB,
        referenceC = referenceC,
        value = value,
        listener = listener
    )


    fun <A> calculatedBy(
        a: ObservableProperty<A>,
        @escaping calculation: (A) -> T
    ): ObservableProperty<T> {
        this.value = calculation(a.value)
        a.onChange.add(listener = weakLambda { aValue ->
            this.value = calculation(aValue)
        })
        return this
    }

    fun <A, B> calculatedBy(
        a: ObservableProperty<A>,
        b: ObservableProperty<B>,
        @escaping calculation: (A, B) -> T
    ): ObservableProperty<T> {
        this.value = calculation(a.value, b.value)
        val aWeak by weak(a)
        val bWeak by weak(b)
        a.onChange.add(listener = weakLambda { aValue ->
            val bValue = bWeak?.value;
            if (bValue != null) {
                this.value = calculation(aValue, bValue)
            }
        })
        b.onChange.add(listener = weakLambda { bValue ->
            val aValue = aWeak?.value;
            if (aValue != null) {
                this.value = calculation(aValue, bValue)
            }
        })
        return this
    }

    fun <A, B, C> calculatedBy(
        a: ObservableProperty<A>,
        b: ObservableProperty<B>,
        c: ObservableProperty<C>,
        @escaping calculation: (A, B, C) -> T
    ): ObservableProperty<T> {
        this.value = calculation(a.value, b.value, c.value)
        val aWeak by weak(a)
        val bWeak by weak(b)
        val cWeak by weak(c)
        a.onChange.add(listener = weakLambda { aValue ->
            val bValue = bWeak?.value;
            val cValue = cWeak?.value;
            if (bValue != null) {
                if (cValue != null) {
                    this.value = calculation(aValue, bValue, cValue)
                }
            }
        })
        b.onChange.add(listener = weakLambda { bValue ->
            val aValue = aWeak?.value;
            val cValue = cWeak?.value;
            if (aValue != null && cValue != null) {
                this.value = calculation(aValue, bValue, cValue)
            }
        })
        c.onChange.add(listener = weakLambda { cValue ->
            val aValue = aWeak?.value;
            val bValue = bWeak?.value;
            if (aValue != null && bValue != null) {
                this.value = calculation(aValue, bValue, cValue)
            }
        })
        return this
    }

    fun <A, B, C, D> calculatedBy(
        a: ObservableProperty<A>,
        b: ObservableProperty<B>,
        c: ObservableProperty<C>,
        d: ObservableProperty<D>,
        @escaping calculation: (A, B, C, D) -> T
    ): ObservableProperty<T> {
        this.value = calculation(a.value, b.value, c.value, d.value)
        val aWeak by weak(a)
        val bWeak by weak(b)
        val cWeak by weak(c)
        val dWeak by weak(d)
        a.onChange.add(listener = weakLambda { aValue ->
            val bValue = bWeak?.value;
            val cValue = cWeak?.value;
            val dValue = dWeak?.value;
            if (bValue != null && cValue != null && dValue != null) {
                this.value = calculation(aValue, bValue, cValue, dValue)
            }
        })
        b.onChange.add(listener = weakLambda { bValue ->
            val aValue = aWeak?.value;
            val cValue = cWeak?.value;
            val dValue = dWeak?.value;
            if (aValue != null && cValue != null && dValue != null) {
                this.value = calculation(aValue, bValue, cValue, dValue)
            }
        })
        c.onChange.add(listener = weakLambda { cValue ->
            val aValue = aWeak?.value;
            val bValue = bWeak?.value;
            val dValue = dWeak?.value;
            if (aValue != null && bValue != null && dValue != null) {
                this.value = calculation(aValue, bValue, cValue, dValue)
            }
        })
        d.onChange.add(listener = weakLambda { dValue ->
            val aValue = aWeak?.value;
            val bValue = bWeak?.value;
            val cValue = cWeak?.value;
            if (aValue != null && bValue != null && cValue != null) {
                this.value = calculation(aValue, bValue, cValue, dValue)
            }
        })
        return this
    }
}
